include/deal.II/lac/petsc_snes.templates.h

```diff
@@ -69,12 +69,16 @@ namespace PETScWrappers
           eptr = nullptr;
           return 0;
         }
-      // In case of an exception, capture the exception and
-      // return -1:
+      catch (const RecoverableUserCallbackError &)
+        {
+          // ok, ignore, but reset the pointer
+          eptr = nullptr;
+          return -1;
```

Comment:
Unless PETSc is actually able to recover, I would rather that `eptr` is set to `std::current_exception`, like for all other exceptions. That's because then once we get back into the deal.II world, we can re-throw that exception to show users what the underlying cause for failure was.